Add tutorial guide for annotated services (#3420)
line/armeria: site/src/pages/tutorials/tutorial-blog-rest-step6.mdx

@Delete("/blogs")
public HttpResponse deleteBlogPost(int id) {
  // Implementation
}
```

## 2. Handle parameters

Let's take the blog post ID (`id`) as a path parameter for identifying the post to delete.

1. Take in the ID value as a path parameter by adding `/blogs/:id` to the <type://@Delete> annotation.
2. [Inject the path parameter](/docs/server-annotated-service#parameter-injection) to the service method by annotating the parameter with <type://@Param>.
  ```java filename=BlogService.java
  @Delete("/blogs/:id")
  public HttpResponse deleteBlogPost(@Param int id) {
    // Implementation
  }
  ```

## 3. Implement service code

In this step, write the code to delete a blog post, handle an exception, and block the operation.

- [Delete a blog post](#delete-a-blog-post)
- [Handle exceptions](#handle-exceptions)
- [Add blocking](#add-blocking)

### Delete a blog post

Deleting a given blog post in this tutorial means removing a blog post from the map, `blogPosts`. However, in real services you would be performing this action on a database.

To delete a blog post, copy the code below to the `deleteBlogPost()` method.

```java filename=BlogService.Java highlight=2

> public HttpResponse deleteBlogPost(@Param int id) {

Reviewer: ditto. Let's keep `@Delete("/blogs/:id")`

Author: Had been added along with adding import statements, 48b7d95931408b1e9f0d0fb877597525ffb443f8

  BlogPost removed = blogPosts.remove(id);
  ...
```

### Handle exceptions

What if there is no such post to delete? We can check if the blog exists before attempting to remove the blog post, but here, let's handle it after the attempt.

1. Throw an `IllegalArgumentException` when no blog post exists with a given ID.
  ```java filename=BlogService.java
  if (removed == null) {
    throw new IllegalArgumentException("The blog post does not exist. id: " + id);
  }
  ```
2. Create an exception handler for the blog service:

>   1. Create a file, BadRequestExceptionHandler.java

Reviewer: ![image](https://user-images.githubusercontent.com/25103250/122151581-c62c2780-ce9a-11eb-9198-1293f86e826f.png)
This does not render properly. We need an extra space. 
```suggestion
   1. Create a file, BadRequestExceptionHandler.java
```

Author: Had been added along with adding import statements, 48b7d95931408b1e9f0d0fb877597525ffb443f8

  ```java filename=BadRequesExceptionHandler.java highlight=3,9,10,12,14
  @Override
  public HttpResponse handleException(ServiceRequestContext ctx, HttpRequest req, Throwable cause) {
    if (cause instanceof IllegalArgumentException) {
        String message = cause.getMessage();
        ObjectNode objectNode = mapper.createObjectNode();
        objectNode.put("error", message);
        try {
            return HttpResponse.of(HttpStatus.BAD_REQUEST,
                                    MediaType.JSON_UTF_8, mapper.writeValueAsString(objectNode));
        } catch (JsonProcessingException e) {
            return HttpResponse.of(HttpStatus.BAD_REQUEST);
        }
    }
    return ExceptionHandlerFunction.fallthrough();
  }
  ```
4. Assign the exception handler to the `deleteBlogPost()` method by annotating the `deletePost()` method with the <type://@ExceptionHandler> as follows.
  ```java filename=BlogService.java highlight=1
  @Delete("/blogs/:id")
  @ExceptionHandler(BadRequestExceptionHandler.class)

>   public HttpResponse deleteBlogPost(@Param int id) {
>     ...

Reviewer: ```suggestion
  public HttpResponse deleteBlogPost(@Param int id) {
    ...
  }
```

Reviewer: or
```suggestion
  public HttpResponse deleteBlogPost(@Param int id) { ... }
```

Author: Opted the second suggestion, thank you.
34b2de0ac1c4ddf269c992308bf94c30353de7ec

  ```

### Add blocking

With real services, accessing and operating on a database takes time. We need to hand over such blocking tasks to [blocking task executor](/docs/server-annotated-service#specifying-a-blocking-task-executor) so that the EventLoop isn't blocked. There are a few options to imeplement this; we'll annotate our service method with the <type://@Blocking>.

```java filename=BlogService.java highlight=1
@Blocking
@Delete("/blogs/:id")
@ExceptionHandler(BadRequestExceptionHandler.class)

> public HttpResponse deleteBlogPost(@Param int id) {

Reviewer: ```suggestion
public HttpResponse deleteBlogPost(@Param int id) { ...}
```

Author: Done 2b82f772c834552d110766b76f22106dba273ccd

  ```java filename=BadRequesExceptionHandler.java highlight=3,9,10,12,14
  @Override
  public HttpResponse handleException(ServiceRequestContext ctx, HttpRequest req, Throwable cause) {
    if (cause instanceof IllegalArgumentException) {
        final String message = cause.getMessage();
        final ObjectNode objectNode = mapper.createObjectNode();
        objectNode.put("error", message);
        try {
            return HttpResponse.of(HttpStatus.BAD_REQUEST,
                                    MediaType.JSON_UTF_8, mapper.writeValueAsString(objectNode));
        } catch (JsonProcessingException e) {
            return HttpResponse.of(HttpStatus.BAD_REQUEST);
        }
    }
    return ExceptionHandlerFunction.fallthrough();
  }
  ```
4. Assign the exception handler to the `deleteBlogPost()` method by annotating the `deletePost()` method with the <type://@ExceptionHandler> as follows.
  ```java filename=BlogService.java highlight=1
  @Delete("/blogs/:id")
  @ExceptionHandler(BadRequestExceptionHandler.class)
  public HttpResponse deleteBlogPost(@Param int id) {
    ...
  ```

### Add blocking

> With real services, accessing and operating on a database takes time. To prevent the service to wait on our task, we want to inform the task executor to handle other tasks until we're done with database operation.

Reviewer: How about
```suggestion
With real services, accessing and operating on a database takes time. We need to hand over those blocking tasks to [blocking task executor](/docs/server-annotated-service#specifying-a-blocking-task-executor) so that the EventLoop isn't blocked.
```

Author: Applied the suggestion, except that I've changed "those" to "such".
Done 264622773062b7bbac0f56eab5f5505fb9e376e9

Reviewer: > I've changed "those" to "such".

That's way better. 😄

---
type: tutorial
tags:
  - rest
  - request
  - request-converter
  - exceptions
  - exception-handling
  - blocking
level: basic
---

# Step 6. Implement DELETE operation

In this step, we will write a method for updating blog posts. Two new things we will try in this section. One is blocking and the other is exception handling. 

// TODO: TW 아래 내용은 다른 형식으로 제공할 수 있음. 고민 중임.

<Tip>

View the sample service code [here](#).

</Tip>

// TODO: TW 이 부분은 별도 섹션이 아닌 adomination으로 변경될 가능성 있음. 고민 중임.
// TODO: List update 예정

After completing this step, you will have learnt to use the following Armeria features:

  - HTTP method mapping (<type://@Delete>)
  - Parameter injection (<type://@Param>)
  - Exception handler (TBU)

>   - Blocking task executor (<type://@Blocking)

Reviewer: ```suggestion
  - Blocking task executor (<type://@Blocking>)
```
Just in case you missed it. 😉

Author: Thank you, :) The list has been moved to somewhere else, but added the closing bracket.

  - exception-handling
  - blocking
level: basic
---

# Implementing DELETE operation

In this step, we'll write a method for deleting a blog post. By completing this step, you'll learn to map your service with the HTTP DELETE (<type://@Delete>) method, customize an exception handler, and use a blocking task executor.

<TutorialSteps current={6} />

## What you need

You must have the following files ready to try creating a blog post. You can always [download](https://github.com/line/armeria-examples/blob/master/tutorials/rest-api-annotated-service/src/main/java/example/armeria/server/blog/) the full version, instead of creating one yourself.

- Main.java: See [Step 1. Create a server](/tutorials/tutorial-blog-rest-step1)
- BlogPost.java: See [Step 2. Prepare a data object](/tutorials/tutorial-blog-rest-step2)

## 0. Create a service file

Create a service file, BlogService.java, if you don't have one yet. Skip this step if you've already made one in earlier steps. You can see the full version of the file [here](https://github.com/line/armeria-examples/blob/master/tutorials/rest-api-annotated-service/src/main/java/example/armeria/server/blog/BlogService.java).

<Warning>

> To test retrieving a blog post, you need to have blog posts created, which requires you to have the creation method implemented in the BlogService.java. See [Step 3. Implement CREATE](/tutorials/tutorial-blog-rest-step3) for instructions.

Reviewer: ```suggestion
To test deleting a blog post, you need to have blog posts created, which requires you to have the creation method implemented in the BlogService.java. See [Step 3. Implement CREATE](/tutorials/tutorial-blog-rest-step3) for instructions.
```
